Standalone - Convert entityTypes to new schema format

colemanw opened the pull request:
Overview
----------------------------------------
EntityTypes conversion.
See  https://lab.civicrm.org/dev/core/-/issues/4999

colemanw committed:
Standalone - Convert entityTypes to new schema format

See  https://lab.civicrm.org/dev/core/-/issues/4999

ufundo commented:
Reviewing this now :+1:

ufundo commented:
Hit some trouble with this.

On a buildkit standalone install, I switched to this branch
```
cd web/core
git checkout origin/pull/30023/head   # also tried origin/pull/30023/merge, same error
cv flush
```

I get 

```

...[CV call stack]...
...[generic flush call stack]....
    1.1244   65969512  25. Civi\Api4\Utils\CoreUtil::getInfoItem($entityName = 'UserJob', $keyToReturn = 'primary_key') /buildkit/build/standalone/web/core/Civi/Api4/Generic/AbstractBatchAction.php:73
    1.1244   65969512  26. Civi\Api4\Provider\ActionObjectProvider->getEntities() /buildkit/build/standalone/web/core/Civi/Api4/Utils/CoreUtil.php:91
    5.1949   68452544  27. Civi\Api4\Generic\AbstractEntity::getInfo() /buildkit/build/standalone/web/core/Civi/Api4/Provider/ActionObjectProvider.php:166
    5.1954   68453800  28. CRM_Core_DAO_Base::fields() /buildkit/build/standalone/web/core/Civi/Api4/Generic/AbstractEntity.php:165
    5.1954   68453800  29. CRM_Core_DAO_Base::getSchemaFields() /buildkit/build/standalone/web/core/CRM/Core/CRM/Core/DAO/Base.php:97
    5.1954   68453800  30. CRM_Core_DAO_Base::loadSchemaFields() /buildkit/build/standalone/web/core/CRM/Core/CRM/Core/DAO/Base.php:106
    5.1957   68461016  31. constant($name = 'CRM_Utils_Type::T_DATE + CRM_Utils_Type::T_TIME') /buildkit/build/standalone/web/core/CRM/Core/CRM/Core/DAO/Base.php:117
```

The offending line is 
```
'type' => !empty($fieldSpec['data_type']) ? \CRM_Utils_Type::getValidTypes()[$fieldSpec['data_type']] : constant(\CRM_Utils_Schema::getCrmTypeFromSqlType($fieldSpec['sql_type'])), 
```

It's weird because I can't see how that return value from  `getCrmTypeFromSqlType` for sql type datetime is ever going to work. Though there are a bunch of other fields with sql type datetime and no data_type set.

colemanw commented:
@artfulrobot @ufundo thanks both for looking at this. I think there are a couple commits in #29991 that would unblock this PR, so let's focus on getting that one merged first. There are currently [a couple puzzling test failures](https://github.com/civicrm/civicrm-core/pull/29991#issuecomment-2106037454) specific to standalone on that PR.

ufundo commented:
So now the failures look like it is trying to double install the `civicrm_uf_match` table.

I think maybe we are being bitten by the fact that the Users entity from standaloneusers is re-using that core table for a separate entity - and so the auto generated equivalent of `auto_install.sql` doesn't work? 

Seems reasonable not to have expected this kind of re-use. I wonder if the simplest way out of there is moving Users to their own `civicrm_standalone_users` table?

colemanw commented on ext/standaloneusers/sql/auto_install.sql:
```diff
@@ -1,113 +0,0 @@
-SET FOREIGN_KEY_CHECKS=0;
-
-DROP TABLE IF EXISTS `civicrm_user_role`;
-DROP TABLE IF EXISTS `civicrm_uf_match`;
```
@ufundo I think this is the missing bit. We could just add this line into the extension's installer somewheres.

ufundo replied:
oh wow ok... so it is currently just nuking any existing uf_match table...


to my mind that seems a bit dangerous? i know currently standaloneusers is the only game in town on standalone, so in practice not a problem, but say you're migrating to standalone from another UF, it would be pref not to nuke that table maybe?

ufundo replied:
just wondering more generally, has the equivalent behaviour to this line been dropped if i install an extension that provides an entity in the new way? was that deliberate?

ufundo replied:
Sorry I know its tangential, but was thinking 'if we don't do this table now, it will be 5 years down the line and someone will be cursing the "clever" reuse of the same table in a different way'.

So I gave it a go: https://github.com/civicrm/civicrm-core/commit/ea35eaf7dc67ef265661ca98d9a74bd9ccc07b44
Thanks to Api4 / this new schema it wasn't so bad... and it seems to install cleanly :)

It slightly replaces one set of exceptions with another - but I do think it makes more sense that Standalone treats the `civicrm_uf_match` table the same way as the other UFs. I feel like we're setting ourselves up for strange bugs otherwise.  E.g. if the entity spec for existing columns in `UFMatch.entityType.php` and `Users.entityType.php` ever diverged....

One thing I couldn't iron out though: I can't `User::get(FALSE)->addSelect(`contact_id.display_name`)` now that contact_id is a "magic" field. It's not terminal, but you lose this column on http://standalone.localhost:7979/civicrm/admin/users. I think it's a general Api4 thing though - it's the same with `Activity::get(FALSE)->addSelect('source_contact_id.display_name')`. Any idea if that's fixable?

ufundo replied:
> "yes, table dropping has been dropped"

Makes a lot of sense to me! The loaded gun analogy :laughing: 

I remember worrying in the past that if you had a table name clash between two extensions, installing one would destroy all the data from the other. 

( Could we one day prefix extension entity table names with their extension name :thought_balloon: )

ufundo replied:
I guess given the substance of table change it probably *should* be a separate PR.

Tried just dropping the core table in the extension installer to get this passing: https://github.com/ufundo/civicrm-core/commit/4b6cd4e83e63a7b54f7d85ec79ee93bb03d303f8

But it doesn't seem to work for me. Trying to call timezone function from the extension before its available by the looks of it... Investigating..

ufundo replied:
Ah, looking at `AutomaticUpgrader` it looks like there's no intended way for the extension to hook in _before_ the table schema creation... so.. not sure what the "easy" answer is

ufundo replied:
Ah, looking at `AutomaticUpgrader` it looks like there's no intended way for the extension to hook in _before_ the table schema creation... so.. not sure what the "easy" fix for this PR is

colemanw replied:
@ufundo I was looking at that too. It looks like this extension was hijacking an internal callback (`onInstall`) which is not really intended to be implemented by extensions, but was the only expedient way to hook in *before* tables are created... but @totten's new automaticUpgrader changes the way these  callbacks work so that now it fires *after* the table is created.

I think I've solved it by adding a new *public* callback for `preInstall`, so that hijacking is no longer necessary.

colemanw committed:
Standalone - drop uf_match table before creating new one

This adds a new preInstall step instead of hijacking the internal onInstall callback.

colemanw commented:
@ufundo this PR effectively eliminates 3 of the 4 drop tables in the installer, so the "glass half full" way to look at it is that it's a net reduction of loaded guns. So perhaps it could be merged on that basis and then consider what to do about the last one. 2 possible solutions would be:
1. Create a separate table for users.
2. Somehow teach the installer to `ALTER TABLE` instead of `DROP+CREATE TABLE`

ufundo commented:
>  a net reduction of loaded guns. So perhaps it could be merged on that basis

yep for sure.

still some tests failing, are they flakes?

colemanw commented:
> @ufundo still some tests failing, are they flakes?

The Drupal10 test are surely unrelated. But I'm not sure what's going on with the `standalone-dev` suite. Would you be able to try running it locally with & without this PR?

ufundo commented:
Sorry @colemanw just picking this up now !

ufundo commented:
Last commit looks great; and it r-runs cleanly through the buildkit install for me.

Trying to re-run those failing upgrade tests cause I think they might be flakes.

Generally thinks it's merge ready but probably @totten should see the AutomaticUpgrader hook change
